config/v1/types.go

```diff
@@ -310,3 +310,74 @@ type DelegatedAuthorization struct {
+	// are in namespaces with labels that match the selector, and are in one of the DomainPatterns.
+	// Defaults to the empty LabelSelector, which matches everything.
+	// A domain MUST be specified in order for namespaceSelector to apply
+	// +optional
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty" protobuf:"bytes,5,opt,name=namespaceSelector"`
+
+	// domainPatterns is an optional list of domains for which these annotations are required.
+	// If domainPatterns is specified and a route is created with a spec.host matching one of the domains,
+	// the route must specify the annotations specified in requiredAnnotations.  If domainPatterns is empty,
+	// the specified annotations are IGNORED.
+	//
+	// The use of wildcards is allowed like this: *.foo.com matches everything under foo.com.
+	// foo.com only matches foo.com, so to cover foo.com and everything under it, you must specify *both*.
+	// +optional
+	DomainPatterns []string `json:"domainPatterns,omitempty"`
+
+	// maxAge is the delta time range in seconds during which hosts are regarded as HSTS hosts.
+	// If set to 0, it negates the effect, and hosts are removed as HSTS hosts.
+	// If set to 0 and includeSubdomains is specified, all subdomains of the host are also removed as HSTS hosts.
+	// maxAge is a time-to-live value, and if this policy is not refreshed on a client, the HSTS
+	// policy will eventually expire on that client.
+	// +required
+	MaxAge MaxAgePolicy `json:"maxage"`
+
+	// preloadPolicy directs the client to include hosts in its host preload list so that
+	// it never needs to do an initial load to get the HSTS header (note that this is not defined
+	// in RFC 6797 and is therefore client implementation-dependent).
+	// +optional
+	PreloadPolicy PreloadPolicy `json:"preloadPolicy,omitempty"`
+
+	// includeSubDomainsPolicy means the HSTS Policy should apply to any subdomains of the host's
+	// domain name.  Thus, for the host bar.foo.com, if includeSubDomainsPolicy was set to RequireIncludeSubDomains:
+	// - the host app.bar.foo.com would inherit the HSTS Policy of bar.foo.com
+	// - the host bar.foo.com would inherit the HSTS Policy of bar.foo.com
+	// - the host foo.com would NOT inherit the HSTS Policy of bar.foo.com
+	// - the host def.foo.com would NOT inherit the HSTS Policy of bar.foo.com
+	// +optional
+	IncludeSubDomainsPolicy IncludeSubDomainsPolicy `json:"includeSubDomainsPolicy,omitempty"`
+}
+
+type MaxAgePolicy struct {
+	// -1 means no opinion
+	// omitempty
+	LargestMaxAge int32 `json:"largestMaxAge,omitempty"`
+	// -1 means no opinion.
+	// To set max-age=0 allows the deletion of an existing HSTS header from a host.  This is a necessary
+	// tool for administrators to quickly correct mistakes.  The minimum `max-age` default is -1 instead of 0.
+	// If it is set to 1 in MaxAgePolicy, then the HSTS policy may not be deleted.
+	// omitempty
+	SmallestMaxAge int32 `json:"smallestMaxAge,omitempty"`
```

Comment:
the doc on this field is very good.  We normally try to avoid pointers because a distinction between zero and nil is painful for API consumers, but your doc here is very convincing.  Do you think a `*int32` provides a better experience and allows the default object (with `nil`) to mean no opinion?

Reply:
Agreed.  Made changes.